### PR TITLE
Support for firefox

### PR DESCRIPTION
**Pull request by NivedhaSenthil**
This PR tries to add basic support for firefox. 

#### To launch taiko with firefox:
- Download and install [firefox nightly](https://www.mozilla.org/en-US/firefox/all/#product-desktop-nightly) 
- Set `TAIKO_BROWSER_PATH` to firefox nightly's executable.

Example: `TAIKO_BROWSER_PATH="/Applications/Firefoxnightly.app/Contents/MacOS/firefox" taiko`

#### Known Issues for now:  
- Highlighting element on action is set to false as overlay domain is not supported
- openTab/closeTab does not work as expected since `New` is not available in firefox yet
- Autofocus on first input field does not happen
- file:// protocol does not emit events
- waitFoNavigation does not wait for network calls and frame loads

### CODE REVIEW

**Comment by zabil** [comment]
Can you also update the documentation and FAQ? For example you can change 

```
It is a node library with a clear and concise API to automate the chrome browser
```

to 

```
It is a node library with a clear and concise API to automate the chromium and firefox browsers
```

Also add a section with the title `Firefox support` to list what does not work and what is different.

File: lib/browserLauncher.js [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+  eventHandler.emit('browserCrashed', new Error(errorMessageForChromeProcessCrash()));
+};
+
+function hasChromeProcessKilled() {
+  return chromeProcess && chromeProcess.killed;
+}
+
+const launchChrome = async (options) => {
+  if (chromeProcess && !chromeProcess.killed) {
+    throw new Error('OpenBrowser cannot be called again as there is a chromium instance open.');
```
Comment (zabil) [inline lib/browserLauncher.js:320]:
```suggestion
    throw new Error('openBrowser cannot be called again as there is a chromium instance open.');
```

File: lib/config.js [hunk 2]
```diff
@@ -8,9 +8,13 @@ const defaultConfig = {
   ignoreSSLErrors: false,
   headful: false,
   criConnectionRetries: process.env.TAIKO_CRI_CONNECTION_RETRIES || 50,
-  highlightOnAction: process.env.TAIKO_HIGHLIGHT_ON_ACTION || 'true',
+  firefox:
+    (process.env.TAIKO_BROWSER_PATH && process.env.TAIKO_BROWSER_PATH.includes('firefox')) || false,
```
Comment (zabil) [inline lib/config.js:12]: Make this check case insensitive

File: docs/layout/partials/header.html [hunk 3]
```diff
@@ -383,6 +385,20 @@ <h2 id="taiko-env-variables">Environment Variables</h2>
   </ul>
 </section>
 
+<section class="taiko-firefox-support">
+  <h2 id="taiko-firefox-support">
+    Experiment Firefox support
```
Comment (zabil) [inline docs/layout/partials/header.html:390]:
```suggestion
    Experimental Firefox support
```

File: lib/browserLauncher.js [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    const os = require('os');
+    const CHROME_PROFILE_PATH = path.join(os.tmpdir(), 'taiko_dev_profile-');
+    temporaryUserDataDir = await mkdtempAsync(CHROME_PROFILE_PATH);
+    args.push(`--user-data-dir=${temporaryUserDataDir}`);
+  }
+  setHeadlessArgs(args, options);
+  return args;
+}
+
+function errorMessageForChromeProcessCrash() {
```
Comment (zabil) [inline lib/browserLauncher.js:291]: Is there something similar for to handle firefox crash?

Comment (NivedhaSenthil) [reply lib/browserLauncher.js]: This method will handle any browser process crash, updated the name accordingly.

File: lib/browserLauncher.js [hunk 5]
```diff
@@ -288,50 +288,50 @@ async function setChromeBrowserArgs(options) {
...
-  if (chromeProcess.exitCode === 0) {
+  if (browserProcess.exitCode === 0) {
     throw new Error(
       'The Browser instance was closed either via `closeBrowser()` call, or it exited for reasons unknown to Taiko. You can try launching a fresh instance using `openBrowser()` or inspect the logs for details of the possible crash.',
     );
   }
-  if (chromeProcess.exitCode === null) {
-    message = `Chrome process with pid ${chromeProcess.pid} exited with signal ${chromeProcess.signalCode}.`;
+  if (browserProcess.exitCode === null) {
+    message = `Chrome process with pid ${browserProcess.pid} exited with signal ${browserProcess.signalCode}.`;
```
Comment (zabil) [inline lib/browserLauncher.js:302]: Can we also make this message generic use `Browser` instead of `Chrome`

**Pull request merged**